src/main/java/org/griddynamics/tic_tac_toe/player/HardBotPlayer.java

```diff
@@ -0,0 +1,130 @@
+package org.griddynamics.tic_tac_toe.player;
+
+import org.griddynamics.tic_tac_toe.Cell;
+import org.griddynamics.tic_tac_toe.Game;
+import org.griddynamics.tic_tac_toe.Grid;
+import org.griddynamics.tic_tac_toe.ui.BotTUI;
+
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.List;
+
+/*
+ * Hard bot implementation
+ */
+public final class HardBotPlayer extends Player {
+
+    private final class FakePlayer extends Player {
+
+        /*
+         * Real player copy constructor
+         */
+        FakePlayer(Grid grid, Player origin) {
+            // Calling parent constructor
+            super(grid, origin.getSign());
+
+            // Making moves
+            for (Cell current : origin.getOccupied()) {
+                this.makeMove(current.getX(), current.getY());
+            }
+        }
+
+        /*
+         * Dummy selectCell implementation
+         */
+        @Override
+        protected Cell selectCell() {
+            return null;
+        }
+    }
+
+    /*
+     * Game reference
+     */
+    private final Game game;
+
+    /*
+     * Index of this player in game's original array
+     */
+    private final int originalIndex;
+
+    /*
+     * Index of next free cell to select for move
+     */
+    private int nextCellToMoveIndex;
+
+    /*
+     * Default constructor
+     */
+    public HardBotPlayer(Grid grid, char sign, Game game, int originalIndex) {
+        super(grid, sign);
+        this.game = game;
+        this.originalIndex = originalIndex;
+    }
+
+    /*
+     * selectFreeCell override
+     */
+    @Override
+    protected Cell selectCell() {
+        // Printing message
+        BotTUI.printHardBotMakeMoveMessage();
+
+        // Going though the tree
+        int temp = func(this.getGrid(),
+             this.game.getPlayers(),
+              this.originalIndex - 1);
+
+        // Returning cell
+        return this.getGrid().getFreeCells().get(nextCellToMoveIndex);
+    }
+
+    private int func(Grid lastGrid,
```

Comment:
I would rename this method as "func" doesn't describe the purpose of this functionality